fix: Update config to be flexible for host

raisedadead opened the pull request:
Currently our config is hardcoded for localhost at a few places, which is okay for development. This PR aims to remove the hardcoding and allow values for `process.env.HOST` , `process.env.PORT` and `process.env.SYNC_PORT` to be configured for working in special cases such as C9 IDE.

Tested changes locally and C9 IDE.

This is advantageous in certain scenarios where contributors may want to work with a Cloud IDE, instead of a local environment.

I am up for a blog post on how I do so if this is accepted.

/cc @QuincyLarson @BerkeleyTrue

raisedadead committed:
fix: Update config to be flexible for host

Greenheart reviewed (commented):
I've not tested this, but I like the idea!

raisedadead commented:
@Greenheart yes, the idea is to provide flexibility while not providing any dedicated support for Cloud IDEs.

I'll create a writeup soon to get this working, if this is accepted. This helps in working from anywhere 😉 ..!

BerkeleyTrue commented on server/middlewares/csp.js:
```diff
@@ -5,7 +5,12 @@ let trusted = [
+  let host = 'http://'
+             + (process.env.HOST || 'localhost')
+             + ':'
+             + (process.env.SYNC_PORT || process.env.PORT || '3000');
```
indenting should be two space.

It may be cleaner to put the save the individual peaces into separate const's and than use a templete string to put it together:

```
trusted = trusted.concat([
     `http://${host}:{port}`,
      ... 
    ]);
```

You would also need to update the websocket url with this

BerkeleyTrue commented on server/boot/a-extendUser.js:
```diff
@@ -77,7 +77,7 @@ module.exports = function(app) {
       subject: 'Welcome to freeCodeCamp!',
       protocol: isDev ? null : 'https',
-      host: isDev ? 'localhost' : 'freecodecamp.com',
+      host: isDev ? (process.env.HOST || 'localhost') : 'freecodecamp.com',
```
Not a big fan of additional conditionals in a ternary. Let's break this up into it's own const's and then use the const in the ternary

BerkeleyTrue commented on gulpfile.js:
```diff
@@ -1,4 +1,5 @@
 // enable debug for gulp
+require('dotenv').load();
 process.env.DEBUG = process.env.DEBUG || 'fcc:*';
```
The require should come after the debug setting. I understand the intent, but the debug should always default to fcc and if the user needs to overwrite it for some reason it should be done on the command line

QuincyLarson commented:
@raisedadead is

raisedadead commented:
@QuincyLarson Yes, this is ready!

BerkeleyTrue commented on server/middlewares/csp.js:
```diff
@@ -4,9 +4,13 @@ let trusted = [
 if (process.env.NODE_ENV !== 'production') {
   trusted = trusted.concat([
-    'ws://localhost:3000'
+    `http://${host}:${port}`,
```
This line can be omitted as it is covered by `self`.

see: http://stackoverflow.com/questions/30280370/how-does-content-security-policy-work

raisedadead replied:
I need one confirmation, while `self` would be able to cover this when the host is `localhost` it will not in be the same in case of most cloud IDEs.

C9's editor / terminal are available at an URL like:
`https://ide.c9.io/<username>/<workspace-name>`

While in execution context the app is only available at:
`https://<worspace-name>-<username>.c9users.io`

I think this is the reason, I had to add this in the first place to get past the errors that I got client side.

I'll re-confirm this soon. Thanks for the comments and QA

BerkeleyTrue replied:
self is determined on the client. If the site is being served at `https://<worspace-name>-<username>.c9users.io` then self will point self to this origin.

BerkeleyTrue replied:
@raisedadead ping

BerkeleyTrue commented:
One more change and this can be merged. I've tested it out locally.

raisedadead commented:
Sorry about the delays, have been AFK since long.

QuincyLarson commented:
@raisedadead did you address @BerkeleyTrue's review? If so please mark it resolved. 

I'll leaving this for @BerkeleyTrue to QA since he's already in the loop.

raisedadead commented:
@QuincyLarson yes, I have addressed the review. I can't dismiss/approve Berkeley's comments, for sake of keeping the history.

@BerkeleyTrue whenever you can this is ready for a merge.

BerkeleyTrue committed:
Merge branch 'staging' into fix/update-config-for-debug-setups

QuincyLarson commented:
@raisedadead awesome - thanks for finishing this! And thanks to @BerkeleyTrue for QA'ing!
